problem/0005-protocol-detection.md

```diff
@@ -0,0 +1,189 @@
+# Protocol Detection Configuration
+
+- Contribution Name: Protocol Detection Configuration
+- Implementation Owner:
+- Start Date: 2020-05-04
+- Target Date:
+- RFC PR:
+- Linkerd Issue:
+- Reviewers: @olix0r @adleong
+
+## Summary
+
+[summary]: #summary
+
+Instead of requiring users to configure the protocol for every port being used
+in an application, Linkerd tries to detect the protocol being used. For
+protocols that the proxy knows about (HTTP) and client speaks first protocols,
+this works pretty well. Unfortunately, for server speaks first protocols, the
+proxy is unable to detect what protocol is being used. This causes delays in
+negotiating new connections and in some cases surprising failure behavior when
+adding Linkerd to new services.
+
+Today, when an application is using an unsupported protocol, the solution is to
+skip either a specific inbound/outbound port or an entire range. When a user
+skips ports on a per-pod basis, the `iptables` rules that redirect incoming and
+outgoing traffic are changed slightly. Any traffic for these skipped ports
+actually skips the proxy. This traffic cannot be mTLS'd and there are no metrics
+available.
+
+## Problem Statement
+
+[problem-statement]: #problem-statement
+
+### What problem are you trying to solve?
+
+Users should be able to get the benefit of lower level protocols (L4) when
+Linkerd is unable to detect the higher level protocol (L7). This allows users to
+get all the benefits they would expect from mTLS, TCP metrics and multi-cluster
+routing. To do this, users need to specify whether protocol detection should be
+skipped on a per-port basis.
+
+### Requirements
+
+- Allow ports to skip the proxy entirely (inbound and outbound), maintaining
```

Comment:
In what situations should the proxy be skipped entirely vs just skipping protocol detection?  Protocol detection should be skipped for known SSF protocols and proxy should be skipped when the proxy is suspected of causing problems (i.e. solely for diagnosing and working around bugs)?

Reply:
We know that there are scenarios where users prefer to skip, for instance, 443, since the proxy adds extremely little value for these streams; especially when startup-ordering issues remain a reality.

Administrators can always disable this feature by adding gatekeeper-style policy enforcements in place to prevent such annotations from being set.

Are there specific concerns?

Reply:
I don't have specific concerns.  These two features have some amount of overlap so I just want to make sure we have a good handle on which situations to use each one.

Reply:
I think it is pretty important to have a "break glass in case of emergency" feature. This'll make it so we can downplay port skipping itself but I think it'll be a valuable tool for getting folks out of an incident.